SRPMs in Copr (#410)
packit/packit.dev: content/posts/copr-srpms.md

---

> title: "Building SRPMs in Copr"

Reviewer: We could link to this blog post from dashboard results :O

date: 2022-03-07T12:57:40+01:00

> ---

Reviewer 2: ```suggestion
weight: 62
---
```

### Introduction

If you use Packit to build RPMs for your upstream code changes, likely,
you have already read about how does Packit build your SRPMs. If not, then just a short recap:
Each time an RPM build is triggered, Packit builds an SRPM and then submits
the created SRPM file to Copr where Copr takes care of building the actual RPMs.

> Since users can modify the behaviour of building SRPMs by defining [actions](https://packit.dev/docs/actions/),

Reviewer: ```suggestion
Since you can modify the behaviour of building SRPMs by defining [actions](https://packit.dev/docs/actions/),
```
to be consistent

this process needs to be run in an isolated environment. For this, we implemented our
[sandboxing mechanism](https://github.com/packit/sandcastle),
which simply runs the provided commands in an Openshift pod freshly created for each build.

#### Problems of the previous workflow for SRPM builds

This is a pretty good-functioning workflow, but it has some downsides which have become more and more annoying
with the growing user base. Because of the resources, we have set limits for Openshift pods running at one time.
This directly affects how many SRPM builds can run in parallel. As a result, when there are too many requests for (S)RPM builds,
some can get stuck in the queue while waiting for other builds to finish. Another inconvenience coming with pods being
always freshly created is copying the needed data into

> and from the pod. This has also cost us some hours of desperate debugging of weird errors.

Reviewer: ```suggestion
and from the pod. This has also cost us some months of desperate debugging of weird errors.
```
sorry
😁

Another disadvantage is that users cannot easily configure dependencies for their actions run during building SRPMs.

> We try to install the dependencies manually on-demand, but of course, this isn't that flexible.

Reviewer: ```suggestion
We have to install the dependencies manually on-demand, which is not flexible.
```

> We have been thinking about improving the process for a long time but have never had any clear conclusion.

Reviewer 2: This is no longer true (we now have the conclusion) so it can't be present perfect tense.
```suggestion
We were thinking about improving the process for a long time but never reached any clear conclusion.
```

we build the SRPMs directly in Copr. We knew that there is a way of building SRPMs in Copr,
but weren't aware of the details and how would this fit our use case.

### Implementation of the Copr SRPMs

After some research of the [Copr custom source method](https://docs.pagure.org/copr.copr/custom_source_method.html), we
decided to give it a try. To make Copr build the SRPMs, Copr needs to be provided
with a script that will prepare the sources used to build an SRPM.
Therefore, we created the `packit prepare-sources` command, which mostly reuses existing code that is run also in the sandbox
workflow. It prepares the specfile, archive and other sources and then moves
them to a separate directory. So with the new implementation,
with each request to run (S)RPM build, Packit sends a dynamically created "script" to Copr that invokes our new command.
Here is what the script can look like:

```
#!/bin/sh

git config --global user.email "hello@packit.dev"
git config --global user.name "Packit"
resultdir=$PWD
packit -d prepare-sources --result-dir "$resultdir" --pr-id 676 --job-config-index 2 https://github.com/packit/ogr

```

> You can see it e.g. tells which pull request should be checked out or which job defined

Reviewer: ```suggestion
You can see that a pull request should be checked out or which job defined
```
just a simplification
